### ISSUE TITLE
Spurious Combinator error with default template 

### ISSUE THREAD

**Issue by emarsden** [ballsteve/xrust#95]
An XSL template that includes the following "pass-through" rule will generate a spurious `Combinator` error in `xslt::from_document`  (using the sample code for the xrust:xslt module, simply with this added to the sample XSLT). 

```
  <xsl:template match='@*|node()'>
    <xsl:copy>
      <xsl:apply-templates select='@*|node()'/>
    </xsl:copy>
  </xsl:template>

```

**Comment by ballsteve** [comment]
Thanks for the bug report. Union operations haven't been tested very thoroughly, so I'll add in this example to the test suite and then look at where the problem is. This will be in the XPath parser, pattern matching and expression transformation.

**Comment by emarsden** [comment]
The error for this example has now changed (using the "dev" branch): I see 

```
thread 'main' panicked at /home/emarsden/.cargo/git/checkouts/xrust-bdc2299a4e8ad91c/65a2b9d/src/xslt.rs:399:47:
index out of bounds: the len is 0 but the index is 0
```

with an input XSLT of 

```
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
  <xsl:template match="@*|node()">
    <xsl:copy>
      <xsl:apply-templates select="@*|node()"/>
    </xsl:copy>
  </xsl:template>
</xsl:stylesheet>
```

**Comment by heaths** [comment]
I tried splitting the templates to avoid the OR, but then `node()` was erring with `undefined function: "name test node"`. Commenting that whole template out now, I get `already borrowed: BorrowMutError1.

My code is the same as in your docs for transforming XML, sans some variable name changes.

**Comment by ballsteve** [comment]
Are you able to add a test case for this into the test suite? Send through a PR

**Comment by ballsteve** [comment]
I've created a new branch for this issue - issue-95. I'll be pushing this to the repo shortly

**Comment by heaths** [comment]
Sorry for the delay. Here's a simple repro for the combinator. I applied a workaround but getting a "union_expr" error in `test_transform_split` that I don't recall getting before. I'll try to see if I can repro it again, but I may have lost some context since it's been a while.

<details>
<summary>Repro code</summary>

```rust
#![allow(dead_code)]

use xrust::{
    parser::xml::parse, transform::context::StaticContextBuilder, trees::smite::RNode,
    xslt::from_document, ErrorKind, Item, Node, SequenceTrait,
};

fn transform(template: &'static str) -> Result<String, xrust::Error> {
    const XML: &str = r#"<?xml version="1.0"?>
        <root>
            <child id="1"/>
            <child Id="2"/>
        </root>
    "#;

    let xmldoc = Item::Node(make_from_str(XML)?);
    let xsldoc = make_from_str(template)?;

    let mut static_content = StaticContextBuilder::new()
        .fetcher(|_| {
            Err(xrust::Error::new(
                ErrorKind::NotImplemented,
                "not implemented",
            ))
        })
        .message(|_| Ok(()))
        .parser(|_| {
            Err(xrust::Error::new(
                ErrorKind::NotImplemented,
                "not implemented",
            ))
        })
        .build();

    let mut ctx = from_document(xsldoc, None, make_from_str, |_| Ok(String::new()))?;
    ctx.context(vec![xmldoc], 0);
    ctx.result_document(RNode::new_document());

    let xml = ctx.evaluate(&mut static_content)?;
    Ok(xml.to_xml())
}

fn make_from_str(s: &str) -> Result<RNode, xrust::Error> {
    let doc = RNode::new_document();
    let _ = parse(doc.clone(), s, None)?;
    Ok(doc)
}

#[test]
fn test_transform_combined() {
    const XSL: &str = r#"
        <xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
            <xsl:output method="xml" version="1.0"/>
            <xsl:template match="node() | @*">
                <xsl:copy>
                    <xsl:apply-templates select="node() | @*"/>
                </xsl:copy>
            </xsl:template>
        </xsl:stylesheet>
    "#;

    let actual = transform(XSL).unwrap();
    println!("{actual}");
}

#[test]
fn test_transform_split() {
    const XSL: &str = r#"
        <xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
            <xsl:output method="xml" version="1.0"/>
            <xsl:template match="node()">
                <xsl:copy>
                    <xsl:apply-templates select="node() | @*"/>
                </xsl:copy>
            </xsl:template>
            <xsl:template match="@*">
                <xsl:copy/>
            </xsl:template>
        </xsl:stylesheet>
    "#;

    let actual = transform(XSL).unwrap();
    println!("{actual}");
}
```

</details>

**Comment by ballsteve** [comment]
The main problem is the union operator in the match pattern. I'm working on that now. Keep track of progress in the issue-95 branch.

**Comment by heaths** [comment]
The `unknown function: "name test node"` appears to be coming from a lone:

```xml
<xsl:apply-templates select="node()"/>
```

Same basic repro as I already pasted, but I commented out `xsl:template` elements. More specifically, my stylesheet is below and applies over a criterion SVG report:

<details><summary>Repro: see BUGBUG line</summary>

```xml
<xsl:stylesheet version="1.0"
  xmlns:svg="http://www.w3.org/2000/svg"
  xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
  <xsl:output method="xml" version="1.0" encoding="utf-8" cdata-section-elements="style"/>
  <xsl:template match="svg:svg">
    <xsl:element name="svg" namespace="http://www.w3.org/2000/svg">
      <xsl:apply-templates select="@*"/>

      <!-- Use @media queries to stylize text for readability in different themes -->
      <xsl:element name="style" namespace="http://www.w3.org/2000/svg">
        <xsl:attribute name="style">text/css</xsl:attribute>
        <xsl:text disable-output-escaping="yes"><![CDATA[
          text { fill: #000000 }
          polyline { fill: none; stroke: #000000 }
          @media (prefers-color-scheme: dark) {
            text { fill: #ffffff }
            polyline { fill: none; stroke: #ffffff }
          }
        ]]></xsl:text>
      </xsl:element>

      <!-- BUGBUG: <xsl:apply-templates select="node()"/> -->
    </xsl:element>
  </xsl:template>

  <!-- Remove the "Input" label that overlaps Y axis labels -->
  <!-- <xsl:template match="svg:text[normalize-space() = 'Input']"/> -->

  <!-- Remove hardcoded colors I want to vary on @media queries -->
  <!-- <xsl:template match="svg:text[@fill = '#000000']/@fill"/>
  <xsl:template match="svg:polyline[@stroke = '#000000']/@stroke"/> -->

  <!-- Split templates because of https://github.com/ballsteve/xrust/issues/95 -->
  <!-- <xsl:template match="node()">
    <xsl:copy>
      <xsl:apply-templates select="@*"/>
      <xsl:apply-templates select="node()"/>
    </xsl:copy>
  </xsl:template>
  <xsl:template match="@*">
    <xsl:copy/>
  </xsl:template> -->
</xsl:stylesheet>
```

</details>

With that commented out, I get an `already borrowed: BorrowMutError` error. The rust code is the same as above.

**Comment by ballsteve** [comment]
The issue-95 branch has the fix and has been promoted as a PR.
